File: src/main/kotlin/com/hoi4utils/clausewitz_parser/Parser.kt
package main.kotlin.com.hoi4utils.clausewitz_parser

import main.kotlin.com.hoi4utils.clausewitz_parser.Node.valueObject
import java.io.File
import java.io.IOException
import java.nio.file.Files

/*
* Parser File new
*/
class Parser {
    private val tokens: Tokenizer
    private var rootNode: Node? = null

    constructor(input: String?) {
        /* EOF */
        var input = input
        input += Token.EOF_INDICATOR
        tokens = Tokenizer(input)
    }

    constructor(f: File) {
        /* get input from file */
        var input: String?
        input = try {
            Files.readString(f.toPath())
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

        /* EOF */input += Token.EOF_INDICATOR
        tokens = Tokenizer(input)
    }

    @Throws(ParserException::class)
    fun parse(): Node {
        val value = parseBlockContent(tokens)

        /*
		 * need to reach up to eof indicator
		 * if last token is '}' this could indicate there was a
		 * missing '{' in the code
		 */if (tokens.peek()!!.type != TokenType.eof) {
            throw ParserException(
                "Input not completely parsed by clausewitz-file parser \n" +
                        "\t\tlast token: " + tokens.peek()!!.value
            )
        }
        rootNode = Node(value)
        return rootNode!!
    }

    @Throws(ParserException::class)
    fun parseBlockContent(tokens: Tokenizer): ArrayList<Node?> {
        val nodes = ArrayList<Node?>()
        while (true) {
            val nextToken = tokens.peek()
            if (nextToken!!.type == TokenType.eof || nextToken.value == "}") {
                break
            }
            try {
                val n = parseNode(tokens)
                // check if n is a comment node
                if (n.nameToken!!.type != TokenType.comment) {
                    nodes.add(n)
                }
            } catch (e: ParserException) {
                throw e
            }
        }
        return nodes
    }

    @Throws(ParserException::class)
    fun parseNode(tokens: Tokenizer): Node {
        val name = tokens.next()

        /* skip comments */if (name!!.type == TokenType.comment) {
            val node = Node()
            node.identifier = name.value
            node.nameToken = name
            return node
        }

        // System.out.println(name.value);
        if (name.type != TokenType.string && name.type != TokenType.symbol && name.type != TokenType.number) {
            throw ParserException(
                "Parser: incorrect token type " + name.type + ", token: " + name + " at index: " + name.start
            )
        }
        var nextToken = tokens.peek()
        if (nextToken!!.type != TokenType.operator || nextToken.value.matches("^[,;}]$".toRegex())) {
            while (nextToken!!.value.matches("^[,;]$".toRegex())) {
                tokens.next()
                nextToken = tokens.peek()
            }

            /* handle escaped characters */
            val nameValue = unescapeCharacters(name, name.value)

            // TODO node value?
            val node = Node()
            node.identifier = nameValue
            node.nameToken = name
            node.operator = null
            node.operatorToken = null
            /* node.value = null; */return node
        }

        /* "= {" */
        val operator: Token?
        if (nextToken.value == "{") {
            operator = Token(
                "=",
                nextToken.start,
                TokenType.operator
            ) // Create a new Token with the '=' value
            nextToken = Token("=", operator.start, TokenType.operator)
            nextToken.start = operator.start
            operator.value = "="
        } else {
            operator = tokens.next()
        }
        var parsedValue: NodeValue? = parseNodeValue(tokens)

        /* Handle value attachment (e.g., when there's a nested block) */if (parsedValue != null && parsedValue.valueObject() is Node) {
            val peekedToken = tokens.peek()
            if (peekedToken!!.value == "{") {
                parsedValue = parseNodeValue(tokens)
            }
        }

        // Skip comments before tailComma
        var tailComma = tokens.peek()
        while (tailComma!!.type == TokenType.comment || tailComma.value.matches("^[,;]".toRegex())) {
            tokens.next()
            tailComma = tokens.peek()
        }

        // todo
        val node: Node = Node(parsedValue)
        node.identifier = name.value
        node.nameToken = name
        node.operator = operator!!.value
        node.operatorToken = operator
        // node.value = parsedValue;
        // node.valueStartToken = (Token) parsedValue[1];
        // node.valueEndToken = (Token) parsedValue[2];
        return node
    }

    private fun unescapeCharacters(name: Token?, nameValue: String?): String? {
        var nameValue = nameValue
        if (name!!.type == TokenType.string) {
            nameValue = nameValue!!.substring(1, nameValue.length - 2)
                .replace(escape_quote_regex.toRegex(), "\"")
                .replace(escape_backslash_regex.toRegex(), "\\")
        }
        return nameValue
    }

    @Throws(ParserException::class)
    fun parseNodeValue(tokens: Tokenizer): NodeValue? {
        val nextToken = tokens.next() ?: return null
        // todo maybe?
        return when (nextToken.type) {
            TokenType.string -> {
                /* substring from 1 to length() - 2: don't replace "" */
                if (nextToken.value.length == 2) {
                    NodeValue(nextToken.value)
                } else NodeValue(
                    nextToken.value.substring(1, nextToken.length() - 2)
                        .replace(
                            escape_quote_regex.toRegex(),
                            "\""
                        )
                        .replace(
                            escape_backslash_regex.toRegex(),
                            "\\"
                        )
                )
            }

            TokenType.number -> {
                /* handles hexadecimal or floating-point/integers */
                NodeValue(
                    if (nextToken.value.startsWith("0x")) nextToken.value.substring(2)
                        .toInt(16) else nextToken.value.toDouble()
                )
            }

            TokenType.symbol -> {
                NodeValue(nextToken.value)
            }

            TokenType.operator -> {
                if (nextToken.value == "{") {
                    val result = parseBlockContent(tokens)
                    val right = tokens.next()
                    if (right!!.value != "}") {
                        throw ParserException("Parser expected a matching \"}\"")
                    }
                    return NodeValue(result)
                }
                break // necessary if addtl. case added, so will keep.
            }

            else -> throw ParserException("Unexpected value: " + nextToken.type)
        }
        throw ParserException("Parser expected a string, number, symbol, or {")
    }

    fun rootNode(): Node? {
        return rootNode
    }

    companion object {
        const val escape_backslash_regex = "\\\\"
        const val escape_quote_regex = "\\\\\""
    }
}

File: src/main/scala/com/hoi4utils/clausewitz_parser/Parser.scala
package com.hoi4utils.clausewitz_parser

import java.io.{File, IOException}
import java.nio.file.Files
import scala.collection.mutable.ListBuffer

/*
 * Parser File new
 */
object Parser {
  val escape_backslash_regex = "\\\\"
  val escape_quote_regex = "\\\\\""
}

class Parser {
  final private var tokens: Tokenizer = _
  private var _rootNode: Node = _

  def this(input: String) = {
    this()

    /* EOF */
    val str = input.concat(Token.EOF_INDICATOR) 
    tokens = new Tokenizer(str)
  }

  def this(file: File) = {
    this({
      new String(Files.readAllBytes(file.toPath))
    })
  }

  @throws[ParserException]
  def parse: Node = {
    val value = parseBlockContent(tokens)
    if (value.isEmpty) {
      throw new ParserException("parsed block content was empty")
    }

    /*
     * need to reach up to eof indicator
     * if last token is '}' this could indicate there was a
     * missing '{' in the code
     */
    tokens.peek match {
      case Some(token) =>
        token.`type` match {
          case TokenType.eof => // good
          case _ => throw new ParserException("Input not completely parsed by clausewitz-file parser \n" + "\t\tlast token: " + token.value)
        }
      case None => throw new ParserException("Input not completely parsed by clausewitz-file parser \n" + "\t\tlast token: null")
    }
    _rootNode = new Node(value)
    _rootNode
  }

  @throws[ParserException]
  def parseBlockContent(tokens: Tokenizer): ListBuffer[Node] = {
    val nodes = new ListBuffer[Node]

    var cont = true
    while (cont) {
      val nextToken: Token = tokens.peek.getOrElse(throw new ParserException("Unexpected null next token"))
      if ((nextToken.`type` eq TokenType.eof) || nextToken.value == "}") {
        cont = false
      } else {
        try {
          val n = parseNode(tokens)
          // check if n is a comment node // todo maybe should be optional?
          if (n.nameToken.`type` ne TokenType.comment) nodes.addOne(n)
        } catch {
          case e: ParserException =>
            throw e
        }
      }
    }
    nodes
  }

  @throws[ParserException]
  def parseNode(tokens: Tokenizer): Node = {
    val name = tokens.next.getOrElse(throw new ParserException("Unexpected null next token"))
    /* skip comments */
    if (name.`type` eq TokenType.comment) {
      val node = new Node
      node.identifier = name.value
      node.nameToken = name
      return node
    }
    // System.out.println(name.value);
    // todo case
    if ((name.`type` ne TokenType.string) && (name.`type` ne TokenType.symbol) && (!name.isNumber))
      throw new ParserException("Parser: incorrect token type " + name.`type` + ", token: " + name + " at index: " + name.start)
    var nextToken = tokens.peek.getOrElse(throw new ParserException("Unexpected null next token"))
    // 'if the next token is not an operator or a [special character]'
    // '^[]': means must contain exactly *one* of the characters within the brackets
    if ((nextToken.`type` ne TokenType.operator) || nextToken.value.matches("^[,;}]$")) {
      /*
      example where you would make it inside here:
        color = { 1.0 1.0 1.0 }
        colortwo = { 1.0 1.0 1.0 }
      each 1.0 will be a node which makes it within here
       */
      while (nextToken.value.matches("^[,;]$")) {
        tokens.next
        nextToken = tokens.peek.getOrElse(throw new ParserException("Unexpected null next token"))
      }
      /* handle escaped characters */
//      val nameValue = unescapeCharacters(name, name.value)
      val parsedValue = parseThisTokenValue(name)

      /* node that is only a value, such as '0.0'. this node has no identifier (no lhs) */
      val node = new Node(parsedValue)
      node.identifier = null
      node.nameToken = name // todo i am lazy this should probably be null. then again everything *should* maybe be redone atp anyways 
      node.operator = null
      node.operatorToken = null
      /* node.value = null; */
      return node
    }

    /* "= {" */
    val operator: Token = {
      if (nextToken.value == "{") {
        // Create a new Token with the '=' value
        val op = new Token("=", nextToken.start, TokenType.operator)
        nextToken = new Token("=", op.start, TokenType.operator)
        op
      } else {
        tokens.next.get
      }
    }

    var parsedValue = parseNodeValue(tokens)
    /* Handle value attachment (e.g., when there's a nested block) */
    if (parsedValue != null && parsedValue.value.isInstanceOf[Node]) {
      val peekedToken = tokens.peek.get
      if (peekedToken.value == "{") {
        parsedValue = parseNodeValue(tokens)
      }
    }
    // Skip comments before tailComma
    var tailComma = tokens.peek.get
    while ((tailComma.`type` eq TokenType.comment) || tailComma.value.matches("^[,;]")) {
      tokens.next
      tailComma = tokens.peek.getOrElse(throw new ParserException("Unexpected null next token"))
    }
    // todo
    val node = new Node(parsedValue)
    node.identifier = name.value
    node.nameToken = name
    node.operator = operator.value
    node.operatorToken = operator
    // node.value = parsedValue;
    // node.valueStartToken = (Token) parsedValue[1];
    // node.valueEndToken = (Token) parsedValue[2];
    node
  }

  private def unescapeCharacters(name: Token, nameValue: String): String = {
    if (name.`type` eq TokenType.string) nameValue
      .substring(1, nameValue.length - 2)
      .replaceAll(Parser.escape_quote_regex, "\"")
      .replaceAll(Parser.escape_backslash_regex, "\\")
    else nameValue
  }

  @throws[ParserException]
  def parseNodeValue(tokens: Tokenizer): NodeValue = {
    val nextToken = tokens.next.getOrElse(return null)
    // todo eeeh?
    nextToken.`type` match {
      case TokenType.string =>
        if (nextToken.length == 1) System.out.println("Parser: ?? " + nextToken.value)
        /* substring from 1 to length() - 2: don't replace "" */
        if (nextToken.value.length == 2) return new NodeValue(nextToken.value)
        return new NodeValue(nextToken.value.substring(1, nextToken.length - 2).replaceAll(Parser.escape_quote_regex, "\"").replaceAll(Parser.escape_backslash_regex, "\\"))

      case TokenType.`float` => return new NodeValue(nextToken.value.toDouble)

      case TokenType.`int` => return new NodeValue(
        if (nextToken.value.startsWith("0x")) Integer.parseInt(nextToken.value.substring(2), 16)
        else nextToken.value.toInt
      )

      case TokenType.symbol => return new NodeValue(nextToken.value)

      case TokenType.operator => if (nextToken.value == "{") {
        val result = parseBlockContent(tokens)
        val right = tokens.next.getOrElse(throw new ParserException("Parser expected a matching \"}\""))
        if (!(right.value == "}")) throw new ParserException("Parser expected a matching \"}\"")
        return new NodeValue(result)
      }

      case _ => throw new ParserException("Unexpected value: " + nextToken.`type`)
    }
    throw new ParserException("Parser expected a string, number, symbol, or {")
  }

  @throws[ParserException]
  def parseThisTokenValue(token: Token): NodeValue = {
    // todo eeeh?
    token.`type` match {
      case TokenType.string =>
        if (token.length == 1) System.out.println("Parser: ?? " + token.value)
        /* substring from 1 to length() - 2: don't replace "" */
        if (token.value.length == 2) return new NodeValue(token.value)
        return new NodeValue(token.value.substring(1, token.length - 2).replaceAll(Parser.escape_quote_regex, "\"").replaceAll(Parser.escape_backslash_regex, "\\"))
      
      case TokenType.`float` => return new NodeValue(token.value.toDouble)
      
      case TokenType.`int` => return new NodeValue(
        if (token.value.startsWith("0x")) Integer.parseInt(token.value.substring(2), 16)
        else token.value.toInt
      )
      
      case TokenType.symbol => return new NodeValue(token.value)
      
      case _ => throw new ParserException("Unexpected value: " + token.`type`)
    }
    throw new ParserException("Parser expected a string, number, or symbol")
  }

  def rootNode: Node = this._rootNode
}
